src/System.IO.FileSystem/src/System/IO/FindEnumerable.Windows.cs

```diff
@@ -0,0 +1,228 @@
+using System.Threading;
+
+namespace System.IO
+{
+    internal unsafe partial class FindEnumerable<TResult, TState> : CriticalFinalizerObject, IEnumerable<TResult>, IEnumerator<TResult>
+    {
+        private readonly string _originalFullPath;
+        private readonly string _originalUserPath;
+        private readonly bool _recursive;
+        private readonly FindTransform<TResult> _transform;
+        private readonly FindPredicate<TState> _predicate;
+        private readonly int _threadId;
+        private readonly TState _state;
+
+        private int _enumeratorCreated;
+
+        private Interop.NtDll.FILE_FULL_DIR_INFORMATION* _info;
+        private byte[] _buffer;
+        private IntPtr _directoryHandle;
+        private string _currentPath;
+        private bool _lastEntryFound;
+        private Queue<(IntPtr Handle, string Path)> _pending;
+        private GCHandle _pinnedBuffer;
+
+        /// <summary>
+        /// Encapsulates a find operation.
+        /// </summary>
+        /// <param name="directory">The directory to search in.</param>
+        public FindEnumerable(
+            string directory,
+            FindTransform<TResult> transform,
+            FindPredicate<TState> predicate,
+            TState state = default,
+            bool recursive = false)
+        {
+            _originalUserPath = directory;
+            _originalFullPath = Path.GetFullPath(directory);
+            _recursive = recursive;
+            _predicate = predicate ?? throw new ArgumentNullException(nameof(predicate));
+            _transform = transform ?? throw new ArgumentNullException(nameof(transform));
+            _threadId = Environment.CurrentManagedThreadId;
+            _state = state;
+            Initialize();
+        }
+
+        private FindEnumerable(
+            string originalUserPath,
+            string originalFullPath,
+            FindTransform<TResult> transform,
+            FindPredicate<TState> predicate,
+            TState state,
+            bool recursive)
+        {
+            _originalUserPath = originalUserPath;
+            _originalFullPath = originalFullPath;
+            _predicate = predicate;
+            _transform = transform;
+            _state = state;
+            _recursive = recursive;
+            _threadId = Environment.CurrentManagedThreadId;
+            Initialize();
+        }
+
+        /// <summary>
+        /// Simple wrapper to allow creating a file handle for an existing directory.
+        /// </summary>
+        public static IntPtr CreateDirectoryHandle(string path)
+        {
+            IntPtr handle = Interop.Kernel32.CreateFile_IntPtr(
+                path,
+                Interop.Kernel32.FileOperations.FILE_LIST_DIRECTORY,
+                FileShare.ReadWrite | FileShare.Delete,
+                FileMode.Open,
+                Interop.Kernel32.FileOperations.FILE_FLAG_BACKUP_SEMANTICS);
+
+            if (handle == IntPtr.Zero || handle == (IntPtr)(-1))
```

Comment:
Do we need to test for IntPtr.Zero here? MSDN says that CreateFile returns -1 on failure.

Reply:
Just was matching the existing SafeHandle. I can remove it.